Add ite-it8xxx2 SoC and board

YCSu0517 opened the pull request:
feat:  add new platformfor ite-it8xxx2 

    We create a new platform under riscv  (boards/riscv/it8xxx2_evb/)
    This commit offer a basic features for the it8xxx2 soc
    More details can reference index.rst

    Signed-off-by: cheryl.su@ite.com.tw <cheryl.su@ite.com.tw>

carlescufi commented:
@YCSu0517  thanks for the contribution. Is this board based on an existing SoC already supported by Zephyr? If not, you should add the new SoC in `soc/`

YCSu0517 commented:
> @YCSu0517 thanks for the contribution. Could you please split this into multiple commits, at the very least:
> 
> * One for the SoC
> * One for the board
> * One for each driver
> 
> Also you shouldn't have commits that amend previous ones. Instead, rebase your set into well-defined units of functionality added.
> 
> More information here:
> https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow

Excuesed me,
I'll splite it, and should I pull multi-requests again? (for different part)
If it.s OK, I'll pull again after fix the errors.
thank you  for reviewing.

jettr commented on boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig:
```diff
@@ -0,0 +1,35 @@
+#CONFIG_RISCV=y
```
Why is this commented out? (It might be correct, just trying to understand)

YCSu0517 replied:
It's seems can't be defined by user?
If opening the define, will cause this error:
error: RISCV (defined at arch/Kconfig:53) is assigned in a configuration file, but is not directly
user-configurable (has no prompt). It gets its value indirectly from other symbols.

So we select this symbol by selecting the ARCH_IS_SET symbol.

jettr commented on drivers/flash/soc_flash_ite_it8xxx2.c:
```diff
@@ -0,0 +1,157 @@
+
+struct flash_priv {
+	/*
+	 * HACK: flash write protection is managed in software.
```
Can you describe this hack more. IS the write_lock never used or something like that?

jettr commented on drivers/flash/soc_flash_ite_it8xxx2.c:
```diff
@@ -0,0 +1,157 @@
+{
+	struct flash_priv *priv = dev->driver_data;
+	uint32_t addr;
+	int rc = -1;
```
variable isn't ever written to and then is used at end for return. This doesn't seem correct

jettr commented on drivers/flash/soc_flash_ite_it8xxx2.c:
```diff
@@ -0,0 +1,157 @@
+	int rc = 0;
+
+	if (enable) {
+		rc = k_sem_take(&priv->write_lock, K_FOREVER);
```
It doesn't appear that this can return anything other than 0 when waiting forever. Might not be worth propagating an error that cannot happen (especially since disable case cannot error)

YCSu0517 replied:
Thanks you for reviewing.
This file has many details need to be modified.
And it cause build error at the latest version zephyr os.
So, We decide to remove it and will pull a new request after rewriting the flash driver.

jettr commented on drivers/serial/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,43 @@
+	  the additional configure options below.
+
+config UART_ITE_IT8XXX2_PORT_0_OPTIONS
+	int "Port 0 Options"
```
What is the key for this options bit field. We should point to documenation

jettr replied:
It has been removed

YCSu0517 replied:
We remove it because it is dummy code on this project.
we use this to verify which port it is on other project.
But in zephyr, the port is treated as different device.

jettr commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+#define t_uart_max_idx          2
+
+enum _T_UART_BAUDRATE_INDEX_ {
+	t_br_4800   = 0,
```
should we drop the =0? Is this enum ever serialized, if not,  let's just let the compile assign values

ElectricalPaul replied:
It does get used as indices to an array (mBaudrateTb[]) later on, so we should probably keep the =0

jettr replied:
Hmm, I still don't think that it worth keeping the =0, but I won't get stuck on this.

jettr commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+{
+	switch (baud_rate) {
+	case 4800:
+		return t_br_4800;
```
If we are going to default to the largest baud rate, should this comparison use >= intead of ==

for example 4801 would produce a baudrate of 115200, I think a better choice would be 4800 or 9600

@carlescufi, Also is there any tracing we can add for these types of error cases that we "silently fix"?

YCSu0517 replied:
Sure, it's more resonable that choice the closer baud rate.
We will rewrite this function:
a. Using call by reference to get baud rate index 
b. The return value will become status for tracing.

carlescufi replied:
@jettr yes, you can use `LOG()` which you should be using anyway to help out in debugging. For hard failures you can also `ASSERT()`.

jettr replied:
This is still unresolved. The code hasn't updated its logic (and it hasn't added LOG statement).

YCSu0517 replied:
We'v updated new code in this block (it became use "if", and will select the >= uart rate), or is it still bad logic? 
and we will add LOG next commit after we check the previous issue.
BR,

jackrosenthal commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,132 @@
+static int ite_intc_init(struct device *dev)
+{
+	irq_connect_dynamic(161, 0,
+		(void (*)(void *))ite_intc_irq_handler, NULL, 0);
```
fix indentation?

jackrosenthal commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,308 @@
+#define TIMER_ENABLE(Idx)               SET_MASK_EXT(IER19, (BIT(3 + (Idx))))
+#define TIMER_CLEAR_FLAG(Idx)       (ISR19 = BIT(3 + (Idx)))
+#define TIMER_CHECK_FLAG(Idx)       (ISR19 & BIT(3 + (Idx)))
+#define TIMER_WAIT(Idx)         (!(ISR19 & BIT(3 + (Idx))))
```
most of these macros can be static functions

trust the compiler does the right thing ;)

jackrosenthal commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,308 @@
+ */
+int timer_init(uint8_t u8Idx,
+	       uint8_t u8PSR, uint8_t bInitialState,
+	       uint8_t bEnableISR, uint32_t u32Cnt)
```
fix indentatoin

jackrosenthal commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,308 @@
+int timer_init_ms(uint8_t u8Idx,
+		  uint8_t u8PSR,
+		  uint8_t bInitialState,
+		  uint8_t bEnableISR, uint32_t u32MilliSec)
```
indentation

jackrosenthal commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,308 @@
+/**
+ * timer_init()
+ */
+int timer_init(uint8_t u8Idx,
```
This variable style is not linux kernel style... generally `lower_case_with_underscores`

jackrosenthal commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,308 @@
+int  timer_count_wait(uint8_t u8Idx)
+{
+	if (u8Idx != EXT_TIMER_8) {
+		while (TIMER_WAIT(u8Idx)) {
```
putting `continue` within the body of the `while` can make it clear

jackrosenthal commented on dts/bindings/interrupt-controller/ite,it8xxx2-intc.yaml:
```diff
@@ -0,0 +1,14 @@
+      required: true
+
+interrupt-cells:
+  - irq
```
a lot of files are missing the newline termination at the end

ElectricalPaul commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,474 @@
+		}
+	}
+}
+static int gpio_ite_pin_interrupt_configure(struct device *dev,
```
Add blank line between functions.

ElectricalPaul commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,474 @@
+static void gpio_ite_isr(void *arg)
+{
+	int i;
+	struct device *dev = arg;
```
Cast void* to target type

andrewboie replied:
why would that be necessary? the C compiler lets you assign void * to other pointer types without complaint

ElectricalPaul replied:
Sorry, you're right. It's C++ needs the explicit cast.

ElectricalPaul commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,474 @@
+
+		}
+	}
+}
```
Add blank line between functions

ElectricalPaul commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,474 @@
+#endif          /* gpio_leds */
+
+
+#define GPIO_ITE_KEY_DEV_CFG_DATA(n) \
```
This macro gets used exactly once, right below where it has been defined. The macro parameter 'n' is used in only one place in the macro. I think it would be easier and clearer to remove the macro and just use it inside the #if below.

ElectricalPaul commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,132 @@
+#include "intc_ite_it8xxx2.h"
+#include <sw_isr_table.h>
+
+static volatile uint8_t *const reg_staus[] = {
```
was this meant to be reg_status?

YCSu0517 replied:
Yes, you are right :)

ElectricalPaul commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,132 @@
+#include "intc_ite_it8xxx2.h"
+#include <sw_isr_table.h>
+
+static volatile uint8_t *const reg_staus[] = {
```
Please define these arrays to have size CONFIG_NUM_IRQS

ElectricalPaul commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,132 @@
+	&ISR20
+};
+
+static volatile uint8_t *const reg_enable[] = {
```
Please define these arrays to have size CONFIG_NUM_IRQS

YCSu0517 replied:
resolved, the length should be MAX_ISR_REG_NUM.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+	unsigned long u32temp = 0;
+
+  #if DEF_ENABLE_CTS_RTS
+	uint8_t Status  = mComPortTb[chan].mUartBase->UMSR;
```
This variable isn't used anywhere in the function.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+	unsigned long u32temp = 0;
+
+  #if DEF_ENABLE_CTS_RTS
+	uint8_t Status  = mComPortTb[chan].mUartBase->UMSR;
```
This variable isn't used anywhere in the function.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+	if (chan == t_uart_1) {
+		GCR21 &= ~(GPH1VS | GPH2VS);
+		SET_MASK(UART2_ECSPMR, BIT1);
+		#if DEF_UART2_ENABLE_CTS_RTS    // {
```
This should be horizontally aligned with the #endif on line 372 below. I would put both in column 1.

ElectricalPaul replied:
Also is there a reason for the "// {" at the end of the line? It looks like it will interfere with brace matching.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+	default:
+		return t_br_115200;
+	}
+	;
```
Stray semicolon should be removed.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+		(volatile struct T_NS16550 *)dev_data->base;
+
+	while ((pComPort->ULSR & LSR_THRE) == 0) {
+	}
```
Should there be a timeout here?

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+		} while (!(ucTemp & MSR_CTS));
+		ucTemp = pComPort->UMSR;
+	}
+    #endif /* DEF_UART2_ENABLE_CTS_RTS */
```
Align with #ifdef above for readability.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+		} while (!(ucTemp & MSR_CTS));
+		ucTemp = pComPort->UMSR;
+	}
+    #endif /* DEF_UART1_ENABLE_CTS_RTS */
```
Align with #ifdef above for readability.

ElectricalPaul commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,959 @@
+{
+	struct uart_ite_it8xxx2_dev_data_t *const dev_data = DEV_DATA(dev);
+	volatile struct T_NS16550 *pComPort =
+		(volatile struct T_NS16550 *)dev_data->base;
```
These three lines get repeated a lot in the functions below. Have you considered declaring an inline function that takes a `struct device*` and returns a `volatile struct T_NS16550*` ?

YCSu0517 replied:
Excuese me, after we checking this, 
We think it couldn't be declare an inline, because the "dev_data" is also be used in the function.

YCSu0517 replied:
The both dev_data and the  pComPort  variable  would be used in partial dunction. (In this case, it's can't be omitted. The two lines are needed)
We remove some dummy code(pComPort) to avoid the repeats.

ElectricalPaul commented on dts/riscv/it8xxx2_evb.dtsi:
```diff
@@ -0,0 +1,112 @@
+				reg = <0x80000000 DT_SIZE_K(512)>;
+				erase-block-size = <1024>;
+				write-block-size = <1>;
+			};
```
Fix indentation for closing brace on this line, line 43, and line 47.

jettr replied:
This for example hasn't been resolved or addressed.


---
Please go through the list of unresolved comments and mark them as resolved when you fix them.

YCSu0517 replied:
OK, I've closed the resolved comment.
By the way, this comment mention the indentation of  closing brace, but I thought it had already been fixed? (If it hasn't, please let me know. Thanks.)

jettr replied:
It looks correct now. Thanks!

ElectricalPaul commented on dts/riscv/it8xxx2_evb.dtsi:
```diff
@@ -0,0 +1,112 @@
+			interrupt-parent = <&intc>;
+		};
+		timer: timer@F01f00 {
+            compatible = "ite,it8xxx2-timer";
```
indentation

ElectricalPaul commented on soc/riscv/riscv-ite/it8xxx2/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+zephyr_sources(
+	soc.c
+
```
delete extra blank line?

ElectricalPaul commented on soc/riscv/riscv-ite/it8xxx2/soc.h:
```diff
@@ -0,0 +1,21 @@
+#define RISCV_RAM_SIZE               KB(CONFIG_SRAM_SIZE)
+
+#define ite_write(reg, reg_size, val)		(*((volatile unsigned char *)(reg))) = val
+#define ite_read(reg, reg_size)				(*((volatile unsigned char *)(reg)))
```
Delete extra spaces to align with the line above

ElectricalPaul commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,2044 @@
+#define BIT0 	0x00000001
+#define BIT1 	0x00000002
+#define BIT2 	0x00000004
+#define BIT3   	0x00000008
```
Delete extra spaces. Either have a consistent number of spaces between the BITx and the 0xvalue (and that should be only 1 space), or line up all the 0xvalues on the same column.

ElectricalPaul commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,2044 @@
+/*
+ * MASK operation macros
+ */
+#define SET_MASK(reg, bit_mask) 	    ((reg)|=(bit_mask))
```
These could be static inline functions.

ElectricalPaul commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,2044 @@
+#define CHECK64KSRAM    ECReg(EC_Register_Base_Address + 0x1060)
+#define CRC_HByte       ECReg(EC_Register_Base_Address + 0x1061)
+#define CRC_LByte       ECReg(EC_Register_Base_Address + 0x1062)
+#define FLHCTRL3R               ECReg(EC_Register_Base_Address + 0x1063)
```
Please align the ECReg columns, for readability.

ElectricalPaul reviewed (commented):
Please see individual comments on files

galak commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,72 @@
+
+Hardware
+********
+The IT8XXX2 series do not support exactly the same hardware features.
```
What is this statement trying to say?  Confused by the 'do not support..'

YCSu0517 replied:
Sorry, It's means that the it8xxxx series have multiple type of mcu, they have different hardware features
for example,
the it81302 is 60 kb sram
the it83202 is 256kb sram
We will rewrite the statement to avoid confusig.

galak reviewed (changes requested):
Can you reduce the board images so the files sizes are not 2+ MB each.

galak commented:
Please fix file mode change of tests/kernel/context/src/main.c.

galak commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,46 @@
+if (SOC_SERIES_RISCV32_IT8XXX2)
+
+config SOC_SERIES
+string
```
Please fix the formatting here,  should be indentation like all other Kconfig files utilize.

galak commented on soc/riscv/riscv-ite/common/soc_common.h:
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2020 ITE Corporation. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
comment here should only be at top of file.

galak commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,2044 @@
+#define CHIP_CHIPREGS_H
+
+#define EC_Register_Base_Address    0x00F00000
+#define BIT_SHIFT(bit, sft) (bit << sft)
```
There are bit macros already defined in include/sys/util.h please use them.

galak commented on dts/bindings/serial/ite,it8xxx2-uart.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2020 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+description: ITE it8xxx2 ns16550 UART
```
If the uart is ns16550 than can we not utilize that existing binding and compatible for that?

YCSu0517 replied:
The it8xxx2 chip is based on ns16550, but there are still some differents.
Such as:
a. The it8xxx2 not support RXRDY and TXRDY functions
b. The it8xxx2 support DMA mode
c. The it8xxx2 have to initial uart module at begin.
...etc.
If binding them, we are afraid of causing the ns16550 code confusion and unreadable.

galak replied:
what do you mean next version?

YCSu0517 replied:
Hi We'v revised our code, it become using the ns16550 seial driver, and removed our uart driver.
The uart hw setting were moved to pinmux.c and soc.c  
Please review it again, thank you.

galak reviewed (commented):
There's already an driver for ns16550 style uarts, can we not utilize and adapt that driver?

ElectricalPaul commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -0,0 +1,121 @@
+            reg = <0x80050000 0x10000>;
+        };
+        storage_partition: partition@80060000 {
+					label = "storage";
```
Please fix indentation to match other partitions above, for readability.

carlescufi commented:
@YCSu0517 please use `git rebase -i` to construct a reasonable set of commits. No "fix" commits at all, split the commits into logical units (one for the SoC, one for each driver, etc) and use `subsystem: title` for the commit message.

jettr commented:
I was just about to make the same comment: please replace the existing CL stack with the ~3 CLs that we would actually want to submit to the master branch. It is difficult to review with the extra fix CLs in the middle and we wouldn't want to submitting it to master like that either.

YCSu0517 commented:
> @YCSu0517 please use `git rebase -i` to construct a reasonable set of commits. No "fix" commits at all, split the commits into logical units (one for the SoC, one for each driver, etc) and use `subsystem: title` for the commit message.

> I was just about to make the same comment: please replace the existing CL stack with the ~3 CLs that we would actually want to submit to the master branch. It is difficult to review with the extra fix CLs in the middle and we wouldn't want to submitting it to master like that either.

Hi, I'v just rebase(rebase to the newest upstream/master) and dropped the old bad commits. 
Instead, spliting them into multi parts, and  update comment for each.
please help to review if it is meet the requirements? 
Thanks all.

jettr commented:
The format of the CLs is much better thank you!

Can you go through and resolve the comments that have been fix and reply to the ones that haven't (with a justification/discussion of why we do not want to implement the suggestion).

As an example, I marked on as resolved that I saw going through the review again, but it would be helpful if you when through and replay to all of the open comments. (Feel free to close a comment if the entire file changed or disappeared and the comment is no longer relevant)

jettr commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,2005 @@
+#define ISR8			ECReg(EC_Register_Base_Address + 0x3F24)
+#define ISR9            ECReg(EC_Register_Base_Address + 0x3F28)
+#define ISR10           ECReg(EC_Register_Base_Address + 0x3F2C)
+#define ISR11           ECReg(EC_Register_Base_Address + 0x3F30)
```
The indentation in this file is off. line 311 uses tabs while line 320 uses spaces. Please be consistent (use tabs please) and make that align

jettr commented:
When reply to comment and making code changes, can you use rebase -i and squash your changes into the original commits instead of adding fix comments to the end of the chain. Then force push your branch again and it will update this PR

jettr commented:
Hey Cheryl, can you please go through this conversation list in the GitHub UI and mark all of the conversations that you have addressed as resolved please? Are there any that are left open?

YCSu0517 commented:
Hi Jettr,

We only opened this conversation currently.
There is no other left, all the resolved issue were shown in this pull-request.
The un-resolved (such as the flash driver or other driver ) issue, we will pull new request after we implement it.

BR, 

> Hey Cheryl, can you please go through this conversation list in the GitHub UI and mark all of the conversations that you have addressed as resolved please? Are there any that are left open?

YCSu0517 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,904 @@
+		}
+#endif  /* DEF_UART2_ENABLE_CTS_RTS */
+	} else if (chan == t_uart_0) {
+#if DEF_UART1_ENABLE_CTS_RTS
```
You are right, this would cause confusion. I revised to use uart1/uart2.

YCSu0517 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,904 @@
+		*baud_rate_index = t_br_4800;
+		if (baud_rate != 4800) {
+			ret = STATUS_WARN_BAUDRATE;
+		}
```
Sure, it's much clear :), done.

jettr commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,424 @@
+static struct gpio_ite_data gpio_ite_data_##n##_##idx; \
+static const struct gpio_ite_cfg gpio_ite_cfg_##n##_##idx = {		\
+	.reg_addr = \
+				DT_REG_ADDR_BY_IDX(DT_PHANDLE_BY_IDX\
```
nit: the formatting seems off here. The indentation doesn't look right.

(This doesn't block PR, but if you are fixing other things, we can probably fix this too)

jettr commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,10 @@
+
+config ITE_IT8XXX2_INTC
+	bool "ITE_IT8XXX2 Interrupt controller"
+	depends on (SOC_IT8XXX2)
```
Should we set a default value here that would work in most cases?

YCSu0517 replied:
Yes, this value is default selected by soc\riscv\riscv-ite\it8xxx2\Kconfig.defconfig.series.

jettr commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,326 @@
+/**
+ * Macro Define
+ */
+#define EXT_CTL_B           (EC_Register_Base_Address + 0x1F10)
```
Where does EC_Register_Base_Address come from? Can you add comments. I would think that this would a relative offset from the base address of the device as specified in the device tree

jettr commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,1943 @@
+
+#include <sys/util.h>
+
+#define EC_Register_Base_Address    0x00F00000
```
I found it, my search wasn't working since this file's diff was hidden...

Shouldn't this value come from the device tree for the board?

YCSu0517 replied:
OK, I've revised it, now it can be defined by devicetree, and the timer register value become come from it's timer device tree.

jettr replied:
Looks good to me, thank you

jettr reviewed (commented):
Seems like this is getting really close. Just a few comments.

I think the biggest of the comments is to try and use the address from the device tree file as the base for all of the relative offsets.

nvlsianpu reviewed (changes requested):
replace usage of zephyr_library_sources_if_kconfig() by zephyr_library_sources_ifdef()
see:
https://github.com/zephyrproject-rtos/zephyr/commit/244f826e3c7333bb92fb53a65c50ee5cbd8a2ea0

nvlsianpu commented:
Also pleas redact commit messages headline in order to reflect where changes were made:
for instance:
`drivers/gpio: it8xxx2 platform driver gpio`
`boards/riscv: add it8xxx2_evb board`
`dts: it8xxx2 device tree and binding`
...

nvlsianpu commented on drivers/serial/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,29 @@
+
+# ---------- Port 0 ----------
+
+config UART_ITE_IT8XXX2_PORT_0
```
Shouldn't a hardware instance be enabled via DT?
@anangl

nvlsianpu replied:
see https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/Kconfig.nrfx#L25

YCSu0517 replied:
I revise it as "def_bool $(dt_nodelabel_has_compat,uart1,it8xxx2-uart)"
The default config will depend on the device tree - uart device. 
(Bythe way, I'v also revised the interrupt config as this way. I am not sure should all the drivers config be enabled via dts?)
BR,

nvlsianpu commented on drivers/serial/Kconfig:
```diff
@@ -128,4 +128,7 @@ source "drivers/serial/Kconfig.xmc4xxx"
 
+source "drivers/serial/Kconfig.it8xxx2"
+
+
```
one new line is enought

nvlsianpu commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Created on: 2015/06/17
```
this info are provided via git commit

nvlsianpu commented:
looks like review fixes are still in separate commits so i overlook that zephyr_library_sources_if_kconfig() was replaced.
Such a fix should be squashed to commit where original code was introduced.
You need to distribute fixing commit among introduction commits.

jettr commented:
Looked over files again. Still looks good to me. Changes have been squash.

@nvlsianpu: Can you take another look, please?

@galak: Can you also take a look and see what we are missing so we can get this merged, please?

andrewboie commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,327 @@
+	return timer_init(idx, psr, initial_state, enable_isr, cnt);
+}
+
+void timer_init_combine(uint8_t idx, uint8_t bEnable)
```
could probably make several of the functions in this file static

YCSu0517 replied:
OK. And I remove some function which were not being used.

andrewboie commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,1944 @@
+
+#include <sys/util.h>
+
+#define EC_Register_Base_Address DT_REG_ADDR_BY_IDX(DT_NODELABEL(baseaddr), 0)
```
would prefer if this was all caps i.e. EC_REGISTER_BASE_ADDRESS (or even EC_REG_BASE_ADDR)

andrewboie commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -0,0 +1,1944 @@
+/*
+ * Macros for hardware registers access.
+ */
+#define ECReg(x)		(*((volatile unsigned char *)(x)))
```
all caps for macro identifiers please

probably should just define in terms of sys_write32, sys_write16, etc

also it seems every usage is `ECReg(EC_Register_Base_Address + <some address>)` why not incorporate the base address into the definition instead of repeating ad nauseam?

YCSu0517 replied:
Hi,

thank for reviewing,
Macro has been fixed to be capital. 
The current ECREG define is flexiable for our patform, so we keep it.
The base address or the register function base address may change with different chip, so we prefer using this way for our chip.

andrewboie commented on soc/riscv/riscv-ite/common/encoding.h:
```diff
@@ -0,0 +1,160 @@
+#ifdef __GNUC__
+
+
+#define read_csr(reg) ({ unsigned long __tmp;				      \
```
not seeing why all these macros in this file have to be #defines instead of inline functions for type safety

YCSu0517 replied:
We found these macros only be used in driver interrupt control.
So, moviing it in to the intc driver and using as inline function.

tyrantlizard77 reviewed (commented):
style nits. substance of the PR looks good

YCSu0517 commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.yaml:
```diff
@@ -0,0 +1,11 @@
+arch: riscv
+toolchain:
+  - zephyr
+ram: 40
```
Yes, it should be 60k(40k is for another chip).
We will revise it.

YCSu0517 commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,243 @@
+
+void z_clock_set_timeout(int32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
```
Yes, and we remove the next if condiction (about idle=null) to avoid confusing.

YCSu0517 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,890 @@
+
+	if (baud_rate <= 4800) {
+		*baud_rate_index = t_br_4800;
+	} else if ((baud_rate <= 9600) && (baud_rate > 4800)) {
```
You are right! It was dummy condiction: ).
We deleted it.

YCSu0517 commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,243 @@
+	if (bEnable) {
+		sys_set_bit((EXT_CTL_B + ((idx >> 1) << (1 + 3))), 3);
+	} else {
+		sys_clear_bit((EXT_CTL_B + ((idx >> 1) << (1 + 3))), 3);
```
Ok, using define replace the idx operations.
#define REG_ADDR_OFFSET(idx)		(idx * MAX_TIMER_NUM)
#define IDX_SHIFT(idx, rsh, lsh)	((idx >> rsh) << lsh)

sjg20 commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,73 @@
+The IT8XXX2 is a 32-bit RISC-V Micro-controller and a
+highly integrated embedded controller with system functions suitable for mobile system applications.
+To find out more about ITE, visit our World Wide Web at:
+`ITE's website <http://www.ite.com.tw>`_.
```
Can you link to the actual product?

YCSu0517 replied:
We have prapared for it.
But it takes some time, we will solved other review first.

sjg20 commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,73 @@
+Overview
+********
+
+The IT8XXX2 is a 32-bit RISC-V Micro-controller and a
```
This line should be longer

sjg20 commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,73 @@
+Listing the IT81302 hardware features as following,
+- RISC-V RV32IMAFC instruction set
+- 4KB instruction cache size
+- Totally 60KB SRAM
```
60KB SDRAM in total

sjg20 commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,73 @@
+In order to upload the application to the device, you'll need our winflash tool and download board.
+
+Then build and flash the application in the usual way.
+   :goals: build flash
```
do you need the 'build' here?

YCSu0517 replied:
yes :)

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+#define GPIO_GPDRM              (EC_Register_Base_Address + 0x160d)
+#define GPCR_OFFSET             (0x10)
+
+struct gpio_ite_wui {
```
Needs comment for struct and members

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+extern void ite_intc_irq_enable(unsigned int irq);
+extern void ite_intc_irq_disable(unsigned int irq);
+
+#define GPIO_GPDRB              (EC_Register_Base_Address + 0x1602)
```
EC_Register_Base_Address should be in CAPs

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+LOG_MODULE_REGISTER(gpio_ite);
+
+extern void ite_intc_irq_enable(unsigned int irq);
```
Function comments

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+	uint32_t pin;
+	uint32_t irq;
+};
+static const struct gpio_ite_wui GPIO_GPDRB_wui[] = {
```
blank lines after }; in each case

What is WUEMR10? A macro?

YCSu0517 replied:
it was register, I've added comment about it.

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+	&IPOLR16, &IPOLR17, &IPOLR18, &IPOLR19,
+	&IPOLR20
+};
+struct gpio_ite_cfg {
```
needs full struct comment. Please comment all structs

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+	((const struct gpio_ite_cfg *)(dev)->config_info)
+
+/**
+ *Helper functions for bit / port access
```
* H

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,420 @@
+		}
+	}
+	/* set wui , only gpiob gpiof, currently */
+	if (gpio_config->reg_addr == GPIO_GPDRB) {
```
This needs a flag in gpio_config as to whether to do this.

Also why are you using GPIO_GPDRB_wui? If it is a macro it should be in all CAPS. But shouldn't you have a pointer in gpio_config abd then use a proper struct? Otherwise this adds to the literal pool

sjg20 commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,13 @@
+
+config ITE_IT8XXX2_INTC
+	bool "ITE_IT8XXX2 Interrupt controller"
+	depends on ( SOC_IT83202|| SOC_IT81302|| SOC_IT8XXX2)
```
drop ()
tidy up spacing: one space is needed in each case

sjg20 commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,13 @@
+	help
+	 Configures the maximum number of clients allowed per shared
+	 instance of the shared interrupt driver. To conserve RAM set
+	 this value to the lowest practical value.
```
Set a default? How do you know what to set this to? I think the help should be more explicit

sjg20 commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,108 @@
+#include <arch/cpu.h>
+#include <init.h>
+#include <sys/printk.h>
+#include "intc_ite_it8xxx2.h"
```
I think this include should go below the next one, to keep the ordering

Also sw comes before sys

sjg20 commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,108 @@
+	&ISR8,  &ISR9,  &ISR10, &ISR11, &ISR12, &ISR13, &ISR14, &ISR15,
+	&ISR16, &ISR17, &ISR18, &ISR19, &ISR20
+};
+static volatile uint8_t *const reg_enable[MAX_ISR_REG_NUM] = {
```
Please fix globally: This needs a blank line between successive decls

sjg20 commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,108 @@
+}
+static int ite_intc_init(struct device *dev)
+{
+	/* INT 161 is software interrupt */
```
add an enum or #define?

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+#define DMACR_TX_FINISH_INT_EN      BIT(5)
+#define DMACR_RX_FINISH_INT_EN      BIT(4)
+#define DMACR_TX_DMA_MODE_EN        BIT(1)
+#define DMACR_RX_DMA_MODE_EN        BIT(0)
```
I think you should use tab instead of lots of spaces

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+ *
+ * uart_init
+ *
+ * @return
```
Drop these comments if they done tell you anything useful. Better to fix them up

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+{
+	int ret = STATUS_SUCCESS;
+
+	if (baud_rate <= 4800) {
```
It seems to me this code would be better as a for loop and a table of baud rates

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+	uint32_t irq_no;
+	#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+	/**< cache of IIR read clear */
```
//*< seems wrong to me since you are not commenting the thing before. I think it should be /**

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+	uint32_t port;
+	uint32_t baud_rate;     /**< Baud rate */
+	uint32_t irq_no;
+	#ifdef CONFIG_UART_INTERRUPT_DRIVEN
```
#ifdef in column 1

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+	volatile struct T_ITEUART *pComPort =
+		(volatile struct T_ITEUART *)dev_data->base;
+
+#if DEF_UART2_ENABLE_CTS_RTS
```
I don't know what DEF_UART2_ENABLE_CTS_RTS is, but I think this should use if() instead of #if. It is better for build coverage and easier to read. The resulting compiler output should be the same. Please fix globally

sjg20 commented on drivers/serial/uart_ite_it8xxx2.c:
```diff
@@ -0,0 +1,931 @@
+	if (dev_data->port == t_uart_1) {
+		do {
+			ucTemp = pComPort->UMSR;
+			ucTemp &= 0xF0;
```
lower-case hex, consistently

sjg20 commented on drivers/timer/Kconfig:
```diff
@@ -169,6 +169,13 @@ config RV32M1_LPTMR_TIMER
+	depends on SOC_IT8XXX2
+	help
+	  This module implements a kernel device driver for the ITE it8xxx2
+	  HW timer model
```
Can you add a few more details? Does it have a configurable frequency? Is it a 16/32-bit timer?...

sjg20 commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,326 @@
+ */
+static void ite_timer_stop(uint8_t idx)
+{
+	ECReg(EXT_CTL_B + ((idx) * 8)) &= ~(BIT(0));
```
EcReg is what? Should be UPPER CASE if a #define

sjg20 commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -0,0 +1,326 @@
+{
+	/* timer_start */
+	ECReg(EXT_CTL_B + ((idx) * 8)) |= BIT(0);
+	ECReg(EXT_LH3R_B + ((idx) * 8)) = (((cnt) >> 24) & 0xFF);
```
Drop () around idx cnt. These are not macros. Please fix globally

sjg20 commented on soc/riscv/riscv-ite/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+config SOC_FAMILY_RISCV_ITE
+	bool
+	# omit prompt to signify a "hidden" option
```
Add help

sjg20 commented on soc/riscv/riscv-ite/common/encoding.h:
```diff
@@ -0,0 +1,161 @@
+#define EXT_IO_BASE        0x40000000
+#define DRAM_BASE          0x80000000
+
+#ifdef __riscv
```
Isn't it always?

sjg20 commented on soc/riscv/riscv-ite/common/encoding.h:
```diff
@@ -0,0 +1,161 @@
+			 __asm__ volatile ("csrr %0, " # reg : "=r" (__tmp)); \
+			 __tmp; })
+
+#define write_csr(reg, val) ({	\
```
These need comments. Can they be inline functions?

sjg20 commented on soc/riscv/riscv-ite/it8xxx2/soc.h:
```diff
@@ -0,0 +1,23 @@
+#define RISCV_RAM_BASE               CONFIG_SRAM_BASE_ADDRESS
+#define RISCV_RAM_SIZE               KB(CONFIG_SRAM_SIZE)
+
+#define ite_write(reg, reg_size, val)\
```
space after )

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -14,19 +14,21 @@
-#define GPIO_HIGH       1
-#define NUM_IO_MAX              8
-#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#define GPIO_LOW	(0)
```
Drop ()

sjg20 commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -14,19 +14,21 @@
+#define GPIO_GPDRB		(DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpiob), 0))
+#define GPIO_GPDRF		(DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpiof), 0))
+#define GPIO_GPDRM		(DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpiom), 0))
+#define GPCR_OFFSET		(0x10)
```
here too

sjg20 commented on drivers/interrupt_controller/intc_ite_it8xxx2.c:
```diff
@@ -29,11 +45,12 @@ static void ite_intc_isr_clear(unsigned int irq)
 	}
-	g = irq / 8;
-	i = irq % 8;
+	g = irq / MAX_REGISR_IRQ_NUM;
```
Shouldn't this commit be squashed into the previous one? I am reviewing the same code twice?

YCSu0517 replied:
I am sorry to have confused you.
It seems like one of the commit cover other commit files. 
I will recommit them to fix this.

YCSu0517 committed:
boards/riscv: add new riscv platform-it8xxx2

We create a new platform for our chip series it8xxx2.
It is a riscv base soc.

Signed-off-by: Cheryl Su <cheryl.su@ite.com.tw>

YCSu0517 committed:
dts: it8xxx2 device tree and binding

This commit is about it8xxx2 platform device tree.
Add driver's binding files, and one device tree as sample.

Signed-off-by: Cheryl Su <cheryl.su@ite.com.tw>

YCSu0517 committed:
drivers/interrupt: it8xxx2 driver interrupt-control

This commit is about platform it8xxx2 interrupt.
The interrupt irqs should reference the datasheet.

Signed-off-by: Cheryl Su <cheryl.su@ite.com.tw>

YCSu0517 committed:
drivers/timer: it8xxx2 platform driver/timer

This commit is about the it8xxx2 timer driver.

We use the timer 5 as system timer for count time,
so the timer interrupt is trigged by it.

Signed-off-by: Cheryl Su <cheryl.su@ite.com.tw>

YCSu0517 committed:
soc/riscv: it8xxx2 soc system

A new platform soc for it8xxx2.
Revising the test/kernel/context/src/main.c for it8xxx2 test case.

Signed-off-by: Cheryl Su <cheryl.su@ite.com.tw>

sjg20 commented:
Is there a way to review changes made in response to previous reviews? Without that I don't think a second round of reviews is all that feasible.

nvlsianpu commented:
> Is there a way to review changes made in response to previous reviews? Without that I don't think a second round of reviews is all that feasible.

No, as we are using sqush in zephyr - but usually this is not a problem.

YCSu0517 commented:
Hi,
Excuese me, we want to know are there any other review requests for this pull-request?
Or could our platform push to the zephyr base?

sjg20 commented:
You have two requested changes. If they haven't replied to your updates you could remove them from the review

YCSu0517 commented:
> Can you reduce the board images so the files sizes are not 2+ MB each.

Hi, I'v reduced the file file, please review it. Thanks a lot.

YCSu0517 commented:
> replace usage of zephyr_library_sources_if_kconfig() by zephyr_library_sources_ifdef()
> see:
> [244f826](https://github.com/zephyrproject-rtos/zephyr/commit/244f826e3c7333bb92fb53a65c50ee5cbd8a2ea0)

Hi , I'v revised them all. Please review it, thanks.

jettr commented:
Friendly pin, @nashif @carlescufi . What is left before this PR can merge?

nvlsianpu commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -0,0 +1,107 @@
+		#size-cells = <1>;
+		boot_partition: partition@80000000 {
+			label = "mcuboot";
+			reg = <0x80000000 0x20000>;
```
Are address ranges here compatible with the flash driver? - All of partition node I saw uses relative addresses (where 0 is flash begin address independently for what the address physically is). This makes flash_map compatible with their flash driver which operates on relative addresses, which make sense as this make API for retrieving flash layout usable. Flash API doesn't provide a way to get address of flash begin.

Summary: flash drivers uses a offsets form flash being, not a physical addresses. On some SoC these value are same - just by chance.  Here same offsets should be used.

YCSu0517 replied:
The address ranges is designed as probable value currently, and it is compatiable. 
The flash driver hasn't completed, this part will be check if we want to pull another request.
And excuse me for asking, are there other examples or platform using offset ? 
I thought this would help to ensure me getting the point.

nvlsianpu replied:
> And excuse me for asking, are there other examples or platform using offset ?
I thought this would help to ensure me getting the point.

All of them. For instance:
from my nRF playground
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf5340dk_nrf5340/nrf5340dk_nrf5340_cpunet.dts
https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nordic/nrf5340_cpunet_qkaa_eng_a.dtsi flash physical addressing begins at 0x01000000.

or STM32F0:
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/stm32f0_disco/stm32f0_disco.dts
https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/st/f0/stm32f051X8.dtsi flash physical addressing begins form 0x08000000

YCSu0517 replied:
Thanks for examples, I understand. :)
The flash physical address has been definded in dtsi, and the flash api get offset from dts.
So I have to revise the partition reg as offset instead of physical address. (Done)

jettr commented:
Anything else left before we can merge?

nvlsianpu commented:
@carlescufi Can be merged.

YCSu0517 commented:
> @carlescufi Can be merged.

Should we do anything else for merging?
There is one change-requested is pending, would it affact the merge?
thanks for help.

nvlsianpu commented:
^^
@galak I think you request were fixed already.
@carlescufi @nashif @MaureenHelm

MaureenHelm reviewed (changes requested):
CI is two months old, please rebase to a more recent master

YCSu0517 commented on dts/bindings/gpio/it8xxx2-gpio.yaml:
```diff
@@ -0,0 +1,23 @@
+
+compatible: "it8xxx2-gpio"
+
+include: [gpio-controller.yaml, base.yaml,gpio-keys.yaml, gpio-leds.yaml ]
```
Hi,
I use keys and led yaml to create devicetree for gpio sample (a button and a led).

MaureenHelm replied:
These compatibles are usually used in the board-level dts, not here

YCSu0517 replied:
The led/key sample is in board-level dts. I''ve got your point,  so i don't need including these yaml in the dts layer:). Thanks, I'll fix this.

YCSu0517 commented on boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig:
```diff
@@ -0,0 +1,28 @@
+CONFIG_SOC_SERIES_RISCV32_IT8XXX2=y
+CONFIG_SOC_IT8XXX2=y
+CONFIG_BOARD_IT8XXX2_EVB=y
+CONFIG_BOOT_DELAY=1
```
Hi, we found partial IC (such as it83201) need delay for hardware initial preparing(or the uart debug message will abnormal). But in some IC, this config can be remove and work well. In order to flexiable, we keep this line.

YCSu0517 commented on drivers/serial/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,29 @@
+
+# ---------- Port 1 ----------
+
+config UART_ITE_IT8XXX2_PORT_1
+	def_bool $(dt_nodelabel_has_compat,uart1,it8xxx2-uart)
+	depends on UART_ITE_IT8XXX2
+	help
+	  This tells the driver to configure the UART port at boot, depending on
+	  the additional configure options below.
+
+# ---------- Port 2 ----------
+
+config UART_ITE_IT8XXX2_PORT_2
```
You are right, I think we can use "ckecking OKAY status" to  replace it.

YCSu0517 commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -0,0 +1,107 @@
+
+/ {
+	model = "IT8XXX2 EV-Board";
+	#compatible = "it8xxx2_evb";
```
Oh, its a mistake. I'll fix it. Thanks.

YCSu0517 committed:
drivers/gpio: it8xxx2 platform gpio driver

This commit is about platform it8xxx2 gpio.
The devicetree use key and led as example.
Users can change it to meet their needs.

Signed-off-by: Cheryl Su <Cheryl.su@ite.com.tw>

YCSu0517 commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,92 @@
+========
+
+In order to upload the application to the device, you'll need our winflash tool and download board.
+Using our winflash.exe tool download the zephyr build flash through our download board to the it8xxx2 board.
```
We'v update our web(http://www.ite.com.tw/en/product/view?mid=149), user can download the datasheet on it. (The winflash tool will be uploaded later.)

MaureenHelm commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,98 @@
+In order to upload the application to the device,
+you'll need our winflash tool and download board.
+You can get the winflash tool at here:`Download Link`_
+Using our winflash.exe tool download the zephyr build flash
+through our download board to the it8xxx2 board.
```
I don't quite understand what this means. I suggest something like this instead:
```
Use the winflash.exe tool to program a zephyr application to the it8xx2 board flash.
```

YCSu0517 replied:
Thanks your suggestion! :)

MaureenHelm commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -0,0 +1,107 @@
+			status = "okay";
+		};
+	};
+	spi0:spi@F02600  {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		compatible = "it8xxx2-sspi";
+		reg = <0x00F02600 0x40>;
+		label = "SPI0";
+		interrupt-parent = <&intc>;
+		interrupts = <37>;
+		clock-frequency = <115200>;
+		status = "okay";
+	};
+	spi1:spi@F02640  {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		compatible = "it8xxx2-sspi";
+		reg = <0x00F02640 0x40>;
+		label = "SPI1";
+		interrupts = <37>;
+		interrupt-parent = <&intc>;
+		status = "okay";
+	};
```
These nodes probably belong in `dts/riscv/it8xxx2_evb.dtsi` instead

YCSu0517 commented on boards/riscv/it8xxx2_evb/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+zephyr_library_include_directories(${PROJECT_SOURCE_DIR}/drivers)
```
This file may be needed in the future, but currently, it  doesn't. I should delete it in this pull-request.

MaureenHelm commented:
@galak please revisit

galak commented on soc/riscv/riscv-ite/it8xxx2/soc.c:
```diff
@@ -0,0 +1,18 @@
+
+#include <device.h>
+#include <init.h>
+#include <sys/util.h>
```
why do we need these headers included?

YCSu0517 replied:
> DTS:
> 
> * Seems like the dts/riscv/it8xxx2_evb.dtsi  should be dts/riscv/it8xxx2.dtsi for the SoC (as I assume EVB means eval board).
> * Its general convention for address to be lowercase (ie `interrupt-controller@F03F00` should be `interrupt-controller@f03f00`
> * looking at the datasheet the INTC supports edge/level and polarity, so those should be encoded in the devicetree.
> 
> Recommend introducing include/dt-bindings/irq.h that has the following defines:
> 
> ```
> #define IRQ_TYPE_NONE           0
> #define IRQ_TYPE_EDGE_RISING    1
> #define IRQ_TYPE_EDGE_FALLING   2
> #define IRQ_TYPE_EDGE_BOTH      (IRQ_TYPE_EDGE_FALLING | IRQ_TYPE_EDGE_RISING)
> #define IRQ_TYPE_LEVEL_HIGH     4
> #define IRQ_TYPE_LEVEL_LOW      8
> ```

Excuese, I try to add this, but there are some issues as follow:
a. I can't find the include/dt-bindings/irq.h file, so I create a file at include/dt-bindings/interrupt-controller/ite-it8xxx2.h. (If this is not appropriate, please let me know.)  
b. I thought the setting intc trigger mode should be coding in arch_irq_priority_set(irq_p, priority_p, flags_p) under riscv/soc, but the function defindef in include/arch/riscv/arch.h is only two parameter. the z_irq_priority_set(irq_p, priority_p, flags_p) will pass three parameters too. Did I make a mistake? Should I coding this at another place?
Thanks you.

galak replied:
> Excuese, I try to add this, but there are some issues as follow:
> a. I can't find the include/dt-bindings/irq.h file, so I create a file at include/dt-bindings/interrupt-controller/ite-it8xxx2.h. (If this is not appropriate, please let me know.)

Might not have been clear, we need to create `include/dt-bindings/irq.h`

> b. I thought the setting intc trigger mode should be coding in arch_irq_priority_set(irq_p, priority_p, flags_p) under riscv/soc, but the function defindef in include/arch/riscv/arch.h is only two parameter. the z_irq_priority_set(irq_p, priority_p, flags_p) will pass three parameters too. Did I make a mistake? Should I coding this at another place?

trigger mode would be handled via flags paramater.

YCSu0517 replied:
Hi,
@galak, @mnkp 
Excuese me for bothering you,  please review again, thank you.

galak commented on boards/riscv/it8xxx2_evb/doc/index.rst:
```diff
@@ -0,0 +1,98 @@
+.. target-notes::
+
+.. _ITE's website: http://www.ite.com.tw/en/product/view?mid=149
+.. _Zephyr Getting Started Guide: https://docs.zephyrproject.org/1.11.0/getting_started/getting_started.html#building-and-running-an-application
```
Pointing to 1.11.0 getting started guide seems wrong.

galak reviewed (changes requested):
DTS:

* Need to add vendor prefix to compatible properties and fix up .dts* and bindings:
* add "ite" to dts/bindings/vendor-prefixes.txt

See other minor comments.

galak commented on dts/riscv/it8xxx2_evb.dtsi:
```diff
@@ -0,0 +1,136 @@
+		#size-cells = <1>;
+		ranges;
+
+		baseaddr: regbase@F00000 {
```
What is this node?  it has no compatible property?

YCSu0517 replied:
It's all of the control register's base address, I think the compatible property should follow the soc? I will add this.

galak replied:
Doesn't seem as if it used anywhere today in the code.

YCSu0517 replied:
This will be used in soc/riscv/riscv-ite/common/chip_chipregs.h .

galak reviewed (changes requested):
DTS:
* Seems like the dts/riscv/it8xxx2_evb.dtsi  should be dts/riscv/it8xxx2.dtsi for the SoC (as I assume EVB means eval board).
* Its general convention for address to be lowercase (ie `interrupt-controller@F03F00` should be `interrupt-controller@f03f00`
* looking at the datasheet the INTC supports edge/level and polarity, so those should be encoded in the devicetree.

Recommend introducing include/dt-bindings/irq.h that has the following defines:
```
#define IRQ_TYPE_NONE           0
#define IRQ_TYPE_EDGE_RISING    1
#define IRQ_TYPE_EDGE_FALLING   2
#define IRQ_TYPE_EDGE_BOTH      (IRQ_TYPE_EDGE_FALLING | IRQ_TYPE_EDGE_RISING)
#define IRQ_TYPE_LEVEL_HIGH     4
#define IRQ_TYPE_LEVEL_LOW      8
```

galak reviewed (commented):
Why does this have a new UART driver, the UART seems to be 16550 compat, can we not use uart_ns16550.c?

mnkp commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,465 @@
+		}
+	} else {
+		ite_write(gpcr_reg_addr + pin, 1, GPIO_DIR_INPUT);
+		if ((flags & GPIO_PULL_DOWN) || (flags & GPIO_PULL_UP)) {
+			return -ENOTSUP;
+		}
```
Let's move this check before configuring the pin as an input.

mnkp commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,465 @@
+	gpcr_reg = ((gpio_config->reg_addr & 0xff) - 1) * NUM_IO_MAX
+			+ gpcr_offset;
+	gpcr_reg_addr = gpcr_reg | (gpio_config->reg_addr & 0xffffff00);
+
```
Let's add
```
	if ((flags & GPIO_SINGLE_ENDED) != 0) {
		return -ENOTSUP;
	}
```
since open-source, open-drain is not supported by the driver.

mnkp commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,465 @@
+#define GPIO_ITE_DEV_CFG_DATA(n, idx, is_out) \
+static struct gpio_ite_data gpio_ite_data_##n##_##idx; \
+static const struct gpio_ite_cfg gpio_ite_cfg_##n##_##idx = { \
+	.reg_addr = DT_REG_ADDR_BY_IDX(DT_PHANDLE_BY_IDX\
+			(DT_NODELABEL(n), gpios, idx), idx),\
```
Let's use `DT_INST_REG_ADDR` macro.

mnkp commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,465 @@
+struct gpio_ite_cfg {
+	gpio_port_pins_t port_pin_mask;
+	uint32_t reg_addr;	/* gpio register base address */
+	char *label;		/* label is set by device tree */
```
There is no need to store label in the config structure. Should be passed directly to `DEVICE_AND_API_INIT` macro.

mnkp commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,465 @@
+		POST_KERNEL, \
+		CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, \
+		&gpio_ite_driver_api)
+#if DT_NODE_HAS_STATUS(DT_INST(0, gpio_leds), okay)
+GPIO_ITE_DEV_CFG_DATA(room_led, 0, 1);
+#endif /* gpio_leds */
+#if DT_NODE_HAS_STATUS(DT_INST(0, gpio_keys), okay)
+GPIO_ITE_DEV_CFG_DATA(room_sw, 0, 0);
+#endif /* gpio_keys */
```
The driver cannot refer to gpio_leds, gpio_keys nodes. This is application specific.

YCSu0517 replied:
It seem's I got them mixed up. I have rewrited this section, I use gpio group as dev init. 
Thanks for your reviewing.

mnkp reviewed (dismissed):
This is a partial review.

mnkp commented on dts/bindings/gpio/it8xxx2-gpio.yaml:
```diff
@@ -0,0 +1,23 @@
+
+description: This binding gives a base representation of the ITE gpio
+
+compatible: "it8xxx2-gpio"
```
@galak This is a new SoC family. Should we require that all compatible properties are prefixed by `ite,`? As in `ite,it8xxx2-gpio` in case of GPIO node.

YCSu0517 replied:
Yes, @galak has reminded me to revised this :) . I'll update this at the next commit.

YCSu0517 committed:
include/de-bindings: create a irq.h for devicetree

This file define the interrupt trigger mode
It's can be used for parsing devicetree
interrupt-control flag setting.

Signed-off-by: Cheryl Su <Cheryl.su@ite.com.tw>

YCSu0517 committed:
drivers/pinmux: it8xxx2 pinmux driver

This driver is setting the it8xxx2 register GCR1-15.
It's for selecting the pin function.
User would set pin at pinmux.c (under board/).

Signed-off-by: Cheryl Su <Cheryl.su@ite.com.tw>

galak commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -0,0 +1,165 @@
+		ranges;
+
+		baseaddr: regbase@f00000 {
+			compatible = "riscv-ite";
```
The compatible here isn't clear.  I'd suggest removing this node for now and adding it back in a future PR when you need it.

galak commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -0,0 +1,165 @@
+		#size-cells = <0>;
+		cpu@0 {
+			clock-frequency = <0>;
+			compatible = "riscv-ite";
```
should this be "ite,riscv-ite"?

YCSu0517 replied:
Done : )

galak commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -0,0 +1,165 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu@0 {
+			clock-frequency = <0>;
```
the `clock-frequency` being 0 seems wrong.  If this should get set in the board dtsi than maybe remove the property here.

YCSu0517 replied:
OK, I'l remov it.  thanks.

galak commented:
Looks like this needs a minor rebase to deal with conflict in CODEOWNERS file

YCSu0517 commented:
> Looks like this needs a minor rebase to deal with conflict in CODEOWNERS file

OK, I'v rebased it and fix the conflict. Thanks.

YCSu0517 commented:
> This is a partial review.

Hi, @mnkp 
Excuese me,
 could you please re-visit this request, thanks.

YCSu0517 commented:
> > This is a partial review.
> 
> Hi, @mnkp
> Excuese me,
> could you please re-visit this request, thanks.
Hi, @mnkp
Sorry for tagging again, but we need to go on this request.
We want to know if there are any oher requests need to be fix?  
thanks.

galak commented on drivers/gpio/gpio_ite_it8xxx2.c:
```diff
@@ -0,0 +1,395 @@
+	.reg_addr = DT_INST_REG_ADDR(idx),\
+	.gpio_irq[0] = DT_INST_IRQ_BY_IDX(idx, 0, irq),	\
+	}; \
+DEVICE_AND_API_INIT(gpio_it8xxx2_dev_##n##_##idx, \
```
Can you look to use `DEVICE_DT_INST_DEFINE` instead of `DEVICE_AND_API_INIT`

galak commented on drivers/pinmux/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,12 @@
+	help
+	  Enable driver for the IT8XXX2 pinmux driver
+
+config PINMUX_ITE_IT8XXX2_NAME
```
Can we add a devicetree node for the pin controller?

galak commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -0,0 +1,105 @@
+	.base = DT_INST_REG_ADDR(0),
+};
+
+DEVICE_AND_API_INIT(pinmux_it8xxx2, CONFIG_PINMUX_ITE_IT8XXX2_NAME,
```
Can we try and use DEVICE_DT_INST_DEFINE or DEVICE_DT_DEFINE instead of DEVICE_AND_API_INIT

YCSu0517 replied:
OK. And we will pull-request for other driver(still coding....) in the future, I think we can update these items at the same time. Thanks!!
